### PR TITLE
grid: Track HTTP/3 broken-ness in ConnectivityGrid.

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Mark HTTP/3 as broken when the HTTP/3 connect attempt fails but the
fallback attempt succeeds.
When HTTP/3 is broken, do not attempt to initiate new HTTP/3 connections.
Add an assert that the Grid's host and the callback's host are the same.

Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
/assign @alyssawilk 

This doesn't yet handle marking QUIC as un-broken and I'm not entire sure I'm in love with this approach for tracking state, so perhaps we should chat about that.

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:92]: Any reason to not create the wrapper callback pointing at the H2 pool if H3 is broken?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:123]: Don't we need to handle the case where H2 succeeds before H3 fails as well?

File: source/common/http/conn_pool_grid.h
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.h:94]: You and your google habits.  no snake case for you. :-P

http3AttemptFailed()
ditto for brokenness below

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:62]: it's a bit hard to tell because things are mocked so much, but I think we can't actually hit this case until we remove the attempt->cancellable_->cancel() from onPoolReady.

Basically right now if H2 connects before H3, we cancel the H3 and it will never actually finish.
so I think we need to amend that cancel call with "if this is H2 connecting, tell the H3 wrapper callback to not do upcalls on completion"

**Review by alyssawilk — COMMENTED** [review]
Sweet!  Excited to see progress on brokenness.  Here's some preliminary thoughts, and happy to sync up when you're in :-)

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h:94]: Done! I renamed is_http3_broken() and set_is_http3_broken(). I *think* that's all?

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: I think we hit this codepath in the case where the HTTP/3 job fails first, and then we start the TCP job. In particular, this happens in ConnectivityGridTest.FailureThenSuccessSerial (added an ASSERT(false) in this code and it fired).

That being said, the case where H2 succeeds before H3 fails is not yet handled.

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h]: Done! I renamed is_http3_broken() and set_is_http3_broken(). I think that's all?

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Sure, that works. Done.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Right, that's the TODO here:

  // TODO: Ensure that if HTTP/2 succeeds, we can allow the HTTP/3 connection to run to completion.                         

How 'bout I address this in a followup since I suspect that might result in a bit of refactoring?

Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Turned out to be less work than I expected, so I've resolved this TODO and handle this case now. (Added a test for it)

**Comment by RyanTheOptimist** [comment]
Blergh, somehow two of my commits did not have signoff on them, do I had to follow the instructions to fix which required a force push. I'd love to understand how those commits did not get signoff.

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:80]: update comment, as pool failure may not be passed up?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:123]: optimization: If H3 connected but H2 was created due to timeout, do we want H2 to complete?  Should we only continue this path if if it's an H3 attempt outstanding?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:134]: optional nit: worth trace logging?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:235]: are you doing this TODO?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.cc:232]: do we need this?  AFIK we can't mark H3 broken until we've already created both pools.

File: source/common/http/conn_pool_grid.h
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.h:95]: nit: the other -> a TCP?

File: source/common/http/conn_pool_grid.h
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.h:115]: Should we call this TCP?  I thought H3 has ALPN for all we hard code it.

File: source/common/http/conn_pool_grid.h
Comment (alyssawilk) [inline source/common/http/conn_pool_grid.h:142]: nit: Envoy rarely does the || style
Also optional Grid's -> grid's?

**Review by alyssawilk — COMMENTED** [review]
Yeah!  This is looking solid :-)
I think you need a main merge, probably to deal with the newStream returns.  Also adding David since it'd be good to have another person familiar with the code, though he's welcome to only do a cursory pass.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc:243]: It's probably not strictly needed. But since the method is a no-op if both pools are already created, it felt a bit "safer" to ensure that we created to pool before increment the pool iterator.  (I could also imagine some future usage in which the Cluster has some a priori knowledge which it uses to mark HTTP/3 broken when creating a ConnectivityGrid).

But I can remove it if you feel strongly.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Sure works for me. I have some feeling in the back of my mind that there might be some corner case where we might wish TCP had run to completion ('cause HTTP/3 is going to end up getting marked as broken) but that doesn't actually seem plausible. Done.

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [reply source/common/http/conn_pool_grid.cc:243]: how about a comment about it likely not being necessary?

File: source/common/http/conn_pool_grid.cc
Comment (alyssawilk) [reply source/common/http/conn_pool_grid.cc:246]: I think you are - by skipping pool ahead above :-P  Functioanlly if QUIC is broken we start at H2

File: test/common/http/conn_pool_grid_test.cc
Comment (alyssawilk) [inline test/common/http/conn_pool_grid_test.cc:239]: should we test the other order as well, of if it's already tested, just verify H3 gets set to broken?

**Review by alyssawilk — COMMENTED** [review]
Looks good!
Just a few more nits and I think we can land this :-)

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc:245]: Done!

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: LOL! I totally misread what the TODO was for. You're right. Removed.

File: test/common/http/conn_pool_grid_test.cc
Comment (RyanTheOptimist) [reply test/common/http/conn_pool_grid_test.cc:239]: I *think* this is what happens in the existing TimeoutThenSuccessParallelSecondConnects test. I think that starts H3, times out, starts TCP, H3 fails (but doesn't up call), TCP succeeds (and upcalls) and we check that H3 is broken.

Does that sound right? Or was there a different ordering you were thinking about?

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks for the review!

**Review by alyssawilk — DISMISSED** [review]
LG!  Merge pending master being unlocked :-)

**Comment by alyssawilk** [comment]
Argh, I was hoping to merge this when the branch opened but we're _still_ slogging through release work and it's not going to happen.
@asraa or @antoniovicente would you mind merging it for me tomorrow since I'll be out?  it's code not used in prod and I've cleared not needing cross-company sign off, just need to merge :-)

**Comment by antoniovicente** [comment]
> Argh, I was hoping to merge this when the branch opened but we're _still_ slogging through release work and it's not going to happen.
> @asraa or @antoniovicente would you mind merging it for me tomorrow since I'll be out? it's code not used in prod and I've cleared not needing cross-company sign off, just need to merge :-)

Sure, I'll look into it tomorrow

**Comment by RyanTheOptimist** [comment]
> > Argh, I was hoping to merge this when the branch opened but we're _still_ slogging through release work and it's not going to happen.
> > @asraa or @antoniovicente would you mind merging it for me tomorrow since I'll be out? it's code not used in prod and I've cleared not needing cross-company sign off, just need to merge :-)
> 
> Sure, I'll look into it tomorrow

Thanks, @antoniovicente !

File: source/common/http/conn_pool_grid.h
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.h:98]: nit: do we define somewhere what "broken" means?

File: source/common/http/conn_pool_grid.h
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.h:117]: nit: it feels weird to have these named "http3" vs "tcp", should we instead say "quic" vs "tcp"?

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:89]: nit: add a log in the else case?

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:113]: you probably already know this, but I'll point it out once because it really surprised me the first time, Envoy `ASSERT` is like `DCHECK`, it's a no-op in release builds.

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:124]: nit: *is* in progress

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:125]: This comment no longer explains *why* we let HTTP/3 complete, should we add a sentence about how it's better so we let it run and if succeeds it'll replace TCP?

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:246]: nit: this is odd, you're incrementing the variable then assigning it to itself? This smells very close to undefined behavior (even though I don't think it is). How about this instead? `pool = std::next(pool);`

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [inline source/common/http/conn_pool_grid.cc:300]: Comparing pointer values can be brittle, any way we could instead add a bool on the pool or something?

File: test/common/http/conn_pool_grid_test.cc
Comment (DavidSchinazi) [inline test/common/http/conn_pool_grid_test.cc:238]: nit: extra dot

**Review by DavidSchinazi — COMMENTED** [review]
Nice!

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h:98]: Added a comment which explains the impact of HTTP/3 being broken which hopefully helps.

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h:118]: "http3" is definitely the right name here because the underlying pool is an "HTTP3" pool, not a QUIC pool (both the name of the class and the type of connections it creates). "tcp" is more debatable. The underlying class is HttpConnPoolImplMixed which we sometimes (I'm told) call "the ALPN pool". I originally used "alpn" instead of "tcp" but alyssa pointed out that HTTP/3 does ALPN so that's not terribly accurate either.

I think "tcp" is the least inaccurate name so I'm inclined to keep it, at least for now.

File: test/common/http/conn_pool_grid_test.cc
Comment (RyanTheOptimist) [reply test/common/http/conn_pool_grid_test.cc]: Wow, I'm surprised that our pedantic spell checking didn't catch that! Done.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc:89]: We already have the log on line 61 that the connection attempt failed. Is that good enough or do you want another log saying that the failure is not being passed up? (I'm inclined to think that the log on line 61 is  but happy to add it if you feel strongly).

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc:113]: Yup! *super* surprising but ok :) (I'd be inclined to rename ASSERT => DEBUG_ASSERT and rename RELEASE_ASSERT => ASSERT, but c'est la vie)

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Done.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Done.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc]: Oh how funny! Not sure how I did that. In any case simply "++pool" will be fine, which I've now done.

File: source/common/http/conn_pool_grid.cc
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.cc:301]: Discussed offline that given the pools are handed out to the connection attempts via pointer it's reasonably safe to use them here. So keeping code as is.

File: source/common/http/conn_pool_grid.cc
Comment (DavidSchinazi) [reply source/common/http/conn_pool_grid.cc:89]: If we have another log then this is fine, thanks for checking!

File: source/common/http/conn_pool_grid.h
Comment (DavidSchinazi) [reply source/common/http/conn_pool_grid.h:98]: It does, thanks!

File: source/common/http/conn_pool_grid.h
Comment (DavidSchinazi) [reply source/common/http/conn_pool_grid.h:118]: I'm not sure that's right. I think of this as a QUIC pool which happens to negotiate HTTP/3 via ALPN. If we were to standardize HTTP/4 which also runs over QUIC (with disambiguation via ALPN like HTTP/1.1 vs HTTP/2) then it would use this pool, not a separate pool. Right?

**Review by DavidSchinazi — COMMENTED** [review]
Thanks!

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h:118]: Hm.  I don't think that QUIC helps make things more readable, specifically because the pool does not vend QUIC streams (the Envoy class StreamEncoder), it vends HTTP/3 streams (the Envoy class RequestEncoder).  The name of the Pool's class is "Http3ConnPoolImpl", not "QuicConnPoolImpl". If at some point this pool vends some other connections than HTTP/3 (say, HTTP/4) we could certainly consider renaming it. But it seems sensible to keep the names consistent in both classes.

That being said, if we feel strongly and want to rename both of them (and perhaps other associated classes) that's definitely an option. Though I'd prefer we do so in a follow up.

Comment (DavidSchinazi) [reply source/common/http/conn_pool_grid.h:118]: Well the TCP pool doesn't vend TCP streams, nor does it vend ALPN streams :P

I'm OK with punting the renaming to a follow-up though, so I'll stop complaining :)

**Review by DavidSchinazi — APPROVED** [review]

File: source/common/http/conn_pool_grid.h
Comment (RyanTheOptimist) [reply source/common/http/conn_pool_grid.h:118]: Definitely up for punting :)

> Well the TCP pool doesn't vend TCP streams, nor does it vend ALPN streams :P

Yes! I totally agree with you here! I think the thing which is poorly named here is "tcp" :) Naming is hard! (Maybe we need to start calling things Legacy HTTP and Modern HTTP :)

**Review by asraa — DISMISSED** [review]

**Review by DavidSchinazi — APPROVED** [review]
Nice fix on the lifetimes!

**Review by asraa — APPROVED** [review]
Thanks! Merging with @alyssawilk out

**Pull request merged**
